File: be/src/statistics/statistics.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository, Between, LessThan, MoreThanOrEqual, LessThanOrEqual } from 'typeorm';
import { Project } from 'src/projects/entities/project.entity';
import { Task } from 'src/tasks/entities/task.entity';
import { ColumnEntity } from 'src/columns/entities/column.entity';
import { Comment } from 'src/comments/entities/comment.entity';
import { User } from 'src/users/entities/user.entity';
import { ProjectOverviewDto } from './dto/project-overview.dto';
import { ColumnStatisticsDto } from './dto/column-statistics.dto';
import { MemberStatisticsDto } from './dto/member-statistics.dto';
import { TimelineStatisticsDto } from './dto/timeline-statistics.dto';
import { CommentStatisticsDto, CommentByTaskDto, CommentByMemberDto } from './dto/comment-statistics.dto';
import { DeadlineAnalyticsDto, TaskDeadlineDto } from './dto/deadline-analytics.dto';

@Injectable()
export class StatisticsService {
  constructor(
    @InjectRepository(Project)
    private readonly projectRepo: Repository<Project>,

    @InjectRepository(Task)
    private readonly taskRepo: Repository<Task>,

    @InjectRepository(ColumnEntity)
    private readonly columnRepo: Repository<ColumnEntity>,

    @InjectRepository(Comment)
    private readonly commentRepo: Repository<Comment>,

    @InjectRepository(User)
    private readonly userRepo: Repository<User>,
  ) {}

  async getProjectOverview(projectId: string): Promise<ProjectOverviewDto> {
    const project = await this.projectRepo.findOne({ where: { id: projectId } });
    if (!project) {
      throw new NotFoundException('Không tìm thấy dự án.');
    }

    const totalColumns = await this.columnRepo.count({
      where: { project: { id: projectId } },
    });

    const tasks = await this.taskRepo
      .createQueryBuilder('task')
      .innerJoin('task.column', 'column')
      .where('column.project_id = :projectId', { projectId })
      .getMany();

    const totalTasks = tasks.length;
    const doneTasks = tasks.filter((t) => t.status === 'done').length;
    const todoTasks = totalTasks - doneTasks;

    const now = new Date();
    const overdueTasks = tasks.filter(
      (t) => t.dueDate && new Date(t.dueDate) < now && t.status !== 'done',
    ).length;

    return {
      totalColumns,
      totalTasks,
      doneTasks,
      todoTasks,
      overdueTasks,
    };
  }

  async getColumnStatistics(projectId: string): Promise<ColumnStatisticsDto[]> {
    const project = await this.projectRepo.findOne({ where: { id: projectId } });
    if (!project) {
      throw new NotFoundException('Không tìm thấy dự án.');
    }

    const columns = await this.columnRepo.find({
      where: { project: { id: projectId } },
      order: { order: 'ASC' },
    });

    const result = await Promise.all(
      columns.map(async (column) => {
        const tasks = await this.taskRepo.find({
          where: { columnId: column.id },
        });

        const totalTasks = tasks.length;
        const doneTasks = tasks.filter((t) => t.status === 'done').length;
        const todoTasks = totalTasks - doneTasks;
        const progress = totalTasks === 0 ? 0 : (doneTasks / totalTasks) * 100;

        return {
          columnId: column.id,
          columnName: column.name,
          totalTasks,
          doneTasks,
          todoTasks,
          progress: Math.round(progress * 100) / 100,
        };
      }),
    );

    return result;
  }

  async getMemberStatistics(projectId: string): Promise<MemberStatisticsDto[]> {
    const project = await this.projectRepo.findOne({ where: { id: projectId } });
    if (!project) {
      throw new NotFoundException('Không tìm thấy dự án.');
    }

    const tasks = await this.taskRepo
      .createQueryBuilder('task')
      .innerJoin('task.column', 'column')
      .leftJoinAndSelect('task.assignees', 'assignees')
      .where('column.project_id = :projectId', { projectId })
      .getMany();

    const userStatsMap = new Map<
      string,
      {
        userId: string;
        name: string;
        avatar: string;
        totalTasks: number;
        doneTasks: number;
        todoTasks: number;
        overdueTasks: number;
      }
    >();

    const now = new Date();

    for (const task of tasks) {
      if (task.assignees && task.assignees.length > 0) {
        for (const assignee of task.assignees) {
          if (!userStatsMap.has(assignee.id)) {
            userStatsMap.set(assignee.id, {
              userId: assignee.id,
              avatar: assignee.avatar || 'https://cdn-icons-png.flaticon.com/512/149/149071.png',
              name: assignee.name || assignee.email,
              totalTasks: 0,
              doneTasks: 0,
              todoTasks: 0,
              overdueTasks: 0,
            });
          }

          const stats = userStatsMap.get(assignee.id)!;
          stats.totalTasks++;

          if (task.status === 'done') {
            stats.doneTasks++;
          } else {
            stats.todoTasks++;
            if (task.dueDate && new Date(task.dueDate) < now) {
              stats.overdueTasks++;
            }
          }
        }
      }
    }

    const result = Array.from(userStatsMap.values()).map((stats) => {
      const completionRate =
        stats.totalTasks === 0 ? 0 : (stats.doneTasks / stats.totalTasks) * 100;
      return {
        ...stats,
        completionRate: Math.round(completionRate * 100) / 100,
      };
    });

    return result.sort((a, b) => b.totalTasks - a.totalTasks);
  }

  async getTimelineStatistics(
    projectId: string,
    period: 'day' | 'week' | 'month' = 'day',
    startDate?: Date,
    endDate?: Date,
  ): Promise<TimelineStatisticsDto[]> {
    const project = await this.projectRepo.findOne({ where: { id: projectId } });
    if (!project) {
      throw new NotFoundException('Không tìm thấy dự án.');
    }

    const tasks = await this.taskRepo
      .createQueryBuilder('task')
      .innerJoin('task.column', 'column')
      .where('column.project_id = :projectId', { projectId })
      .getMany();

    const dateMap = new Map<string, TimelineStatisticsDto>();

    const now = startDate || new Date();
    const end = endDate || new Date();

    for (const task of tasks) {
      const createdDate = new Date(task.createdAt);
      const completedDate = task.completedAt ? new Date(task.completedAt) : null;
      const dueDate = task.dueDate ? new Date(task.dueDate) : null;

      let createdKey: string;
      let completedKey: string | null = null;

      if (period === 'day') {
        createdKey = createdDate.toISOString().split('T')[0];
        if (completedDate) {
          completedKey = completedDate.toISOString().split('T')[0];
        }
      } else if (period === 'week') {
        const weekStart = new Date(createdDate);
        weekStart.setDate(weekStart.getDate() - weekStart.getDay());
        createdKey = weekStart.toISOString().split('T')[0];
        if (completedDate) {
          const weekStartCompleted = new Date(completedDate);
          weekStartCompleted.setDate(weekStartCompleted.getDate() - weekStartCompleted.getDay());
          completedKey = weekStartCompleted.toISOString().split('T')[0];
        }
      } else {
        createdKey = `${createdDate.getFullYear()}-${String(createdDate.getMonth() + 1).padStart(2, '0')}`;
        if (completedDate) {
          completedKey = `${completedDate.getFullYear()}-${String(completedDate.getMonth() + 1).padStart(2, '0')}`;
        }
      }

      if (!dateMap.has(createdKey)) {
        dateMap.set(createdKey, {
          date: createdKey,
          createdTasks: 0,
          completedTasks: 0,
          onTimeTasks: 0,
          lateTasks: 0,
        });
      }
      dateMap.get(createdKey)!.createdTasks++;

      if (completedKey && task.status === 'done') {
        if (!dateMap.has(completedKey)) {
          dateMap.set(completedKey, {
            date: completedKey,
            createdTasks: 0,
            completedTasks: 0,
            onTimeTasks: 0,
            lateTasks: 0,
          });
        }
        const stats = dateMap.get(completedKey)!;
        stats.completedTasks++;

        if (dueDate && completedDate) {
          if (completedDate <= dueDate) {
            stats.onTimeTasks++;
          } else {
            stats.lateTasks++;
          }
        }
      }
    }

    const result = Array.from(dateMap.values()).sort((a, b) => a.date.localeCompare(b.date));
    return result;
  }

  async getCommentStatistics(
    projectId: string,
    filter?: '24h' | '7d' | 'all',
  ): Promise<CommentStatisticsDto> {
    const project = await this.projectRepo.findOne({ where: { id: projectId } });
    if (!project) {
      throw new NotFoundException('Không tìm thấy dự án.');
    }

    const tasks = await this.taskRepo
      .createQueryBuilder('task')
      .innerJoin('task.column', 'column')
      .where('column.project_id = :projectId', { projectId })
      .getMany();

    const taskIds = tasks.map((t) => t.id);

    if (taskIds.length === 0) {
      return {
        totalComments: 0,
        recentComments: 0,
        commentsByTask: [],
        commentsByMember: [],
      };
    }

    let commentsQuery = this.commentRepo
      .createQueryBuilder('comment')
      .leftJoinAndSelect('comment.user', 'user')
      .leftJoinAndSelect('comment.task', 'task')
      .where('comment.taskId IN (:...taskIds)', { taskIds });

    if (filter === '24h') {
      const yesterday = new Date();
      yesterday.setDate(yesterday.getDate() - 1);
      commentsQuery = commentsQuery.andWhere('comment.createdAt >= :yesterday', { yesterday });
    } else if (filter === '7d') {
      const sevenDaysAgo = new Date();
      sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 7);
      commentsQuery = commentsQuery.andWhere('comment.createdAt >= :sevenDaysAgo', {
        sevenDaysAgo,
      });
    }

    const comments = await commentsQuery.getMany();

    const totalComments = comments.length;

    let recentComments = 0;
    if (filter === '24h') {
      const yesterday = new Date();
      yesterday.setDate(yesterday.getDate() - 1);
      recentComments = comments.filter((c) => new Date(c.createdAt) >= yesterday).length;
    } else if (filter === '7d') {
      const sevenDaysAgo = new Date();
      sevenDaysAgo.setDate(sevenDaysAgo.getDate() - 7);
      recentComments = comments.filter((c) => new Date(c.createdAt) >= sevenDaysAgo).length;
    } else {
      recentComments = totalComments;
    }

    const commentsByTaskMap = new Map<string, { taskId: string; taskTitle: string; count: number }>();
    const commentsByMemberMap = new Map<
      string,
      { userId: string; userName: string; avatar: string; count: number }
    >();

    for (const comment of comments) {
      if (comment.task) {
        const taskId = comment.task.id;
        if (!commentsByTaskMap.has(taskId)) {
          commentsByTaskMap.set(taskId, {
            taskId,
            taskTitle: comment.task.title || 'Untitled',
            count: 0,
          });
        }
        commentsByTaskMap.get(taskId)!.count++;
      }

      if (comment.user) {
        const userId = comment.user.id;
        if (!commentsByMemberMap.has(userId)) {
          commentsByMemberMap.set(userId, {
            userId,
            userName: comment.user.name || comment.user.email,
            avatar: comment.user.avatar || 'https://cdn-icons-png.flaticon.com/512/149/149071.png',
            count: 0,
          });
        }
        commentsByMemberMap.get(userId)!.count++;
      }
    }

    const commentsByTask: CommentByTaskDto[] = Array.from(commentsByTaskMap.values())
      .sort((a, b) => b.count - a.count)
      .slice(0, 10)
      .map((item) => ({
        taskId: item.taskId,
        taskTitle: item.taskTitle,
        commentCount: item.count,
      }));

    const commentsByMember: CommentByMemberDto[] = Array.from(commentsByMemberMap.values())
      .sort((a, b) => b.count - a.count)
      .slice(0, 10)
      .map((item) => ({
        userId: item.userId,
        userName: item.userName,
        avatar: item.avatar,
        commentCount: item.count,
      }));

    return {
      totalComments,
      recentComments,
      commentsByTask,
      commentsByMember,
    };
  }

  async getDeadlineAnalytics(projectId: string): Promise<DeadlineAnalyticsDto> {
    const project = await this.projectRepo.findOne({ where: { id: projectId } });
    if (!project) {
      throw new NotFoundException('Không tìm thấy dự án.');
    }

    const tasks = await this.taskRepo
      .createQueryBuilder('task')
      .innerJoin('task.column', 'column')
      .where('column.project_id = :projectId', { projectId })
      .andWhere('task.dueDate IS NOT NULL')
      .getMany();

    const now = new Date();
    const threeDaysLater = new Date(now);
    threeDaysLater.setDate(threeDaysLater.getDate() + 3);

    const overdueTasksList: TaskDeadlineDto[] = [];
    const dueSoonTasksList: TaskDeadlineDto[] = [];
    const completedOnTimeList: TaskDeadlineDto[] = [];
    const completedLateList: TaskDeadlineDto[] = [];

    for (const task of tasks) {
      if (!task.dueDate) continue;

      const dueDate = new Date(task.dueDate);
      const isOverdue = dueDate < now && task.status !== 'done';
      const isDueSoon =
        dueDate >= now && dueDate <= threeDaysLater && task.status !== 'done';

      const taskDto: TaskDeadlineDto = {
        taskId: task.id,
        taskTitle: task.title,
        dueDate: task.dueDate,
        status: task.status,
        completedAt: task.completedAt || null,
      };

      if (isOverdue) {
        overdueTasksList.push(taskDto);
      } else if (isDueSoon) {
        dueSoonTasksList.push(taskDto);
      } else if (task.status === 'done' && task.completedAt) {
        const completedAt = new Date(task.completedAt);
        if (completedAt <= dueDate) {
          completedOnTimeList.push(taskDto);
        } else {
          completedLateList.push(taskDto);
        }
      }
    }

    return {
      overdueTasks: overdueTasksList.length,
      dueSoonTasks: dueSoonTasksList.length,
      completedOnTime: completedOnTimeList.length,
      completedLate: completedLateList.length,
      overdueTasksList,
      dueSoonTasksList,
      completedOnTimeList,
      completedLateList,
    };
  }
}

File: backend-java/src/main/java/com/qlda/backendjava/statistics/service/StatisticsService.java
package com.qlda.backendjava.statistics.service;

import com.qlda.backendjava.common.exception.ForbiddenException;
import com.qlda.backendjava.common.exception.NotFoundException;
import com.qlda.backendjava.columns.repository.ColumnRepository;
import com.qlda.backendjava.comments.entity.CommentEntity;
import com.qlda.backendjava.comments.repository.CommentRepository;
import com.qlda.backendjava.projectmember.entity.ProjectMemberEntity;
import com.qlda.backendjava.projectmember.repository.ProjectMemberRepository;
import com.qlda.backendjava.projects.entity.ProjectEntity;
import com.qlda.backendjava.projects.repository.ProjectRepository;
import com.qlda.backendjava.statistics.dto.*;
import com.qlda.backendjava.tasks.entity.TaskEntity;
import com.qlda.backendjava.tasks.repository.TaskRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.data.jpa.domain.Specification;
import org.springframework.stereotype.Service;

import java.time.LocalDateTime;
import java.time.format.DateTimeFormatter;
import java.time.DayOfWeek;
import java.time.temporal.TemporalAdjusters;
import java.util.*;
import java.util.stream.Collectors;

@Service
@RequiredArgsConstructor
public class StatisticsService {

    private final ProjectRepository projectRepository;
    private final TaskRepository taskRepository;
    private final ColumnRepository columnRepository;
    private final CommentRepository commentRepository;
    private final ProjectMemberRepository projectMemberRepository;

    public ProjectOverviewDto getProjectOverview(String projectId, String userId) {
        ensureUserIsProjectMember(projectId, userId);
        
        ProjectEntity project = projectRepository.findById(projectId)
                .orElseThrow(() -> new NotFoundException("Không tìm thấy dự án."));

        long totalColumns = columnRepository.findByProjectId(projectId).size();

        
        List<com.qlda.backendjava.columns.entity.ColumnEntity> columns = columnRepository.findByProjectId(projectId);
        List<String> columnIds = columns.stream()
                .map(com.qlda.backendjava.columns.entity.ColumnEntity::getId)
                .collect(Collectors.toList());

        
        List<TaskEntity> tasks = new ArrayList<>();
        for (String columnId : columnIds) {
            tasks.addAll(taskRepository.findByColumnIdOrderByPositionAsc(columnId));
        }

        int totalTasks = tasks.size();
        long doneTasks = tasks.stream()
                .filter(t -> t.getStatus() == TaskEntity.Status.done)
                .count();
        int todoTasks = totalTasks - (int) doneTasks;

        LocalDateTime now = LocalDateTime.now();
        long overdueTasks = tasks.stream()
                .filter(t -> t.getDueDate() != null &&
                           t.getDueDate().isBefore(now) &&
                           t.getStatus() != TaskEntity.Status.done)
                .count();

        return new ProjectOverviewDto(
                (int) totalColumns,
                totalTasks,
                (int) doneTasks,
                todoTasks,
                (int) overdueTasks
        );
    }

    public List<ColumnStatisticsDto> getColumnStatistics(String projectId, String userId) {
        ensureUserIsProjectMember(projectId, userId);

        List<com.qlda.backendjava.columns.entity.ColumnEntity> columns = columnRepository.findByProjectId(projectId);

        return columns.stream()
                .map(column -> {
                    List<TaskEntity> tasks = taskRepository.findByColumnIdOrderByPositionAsc(column.getId());
                    int totalTasks = tasks.size();
                    long doneTasks = tasks.stream()
                            .filter(t -> t.getStatus() == TaskEntity.Status.done)
                            .count();
                    int todoTasks = totalTasks - (int) doneTasks;
                    double progress = totalTasks == 0 ? 0 : (doneTasks * 100.0 / totalTasks);

                    return new ColumnStatisticsDto(
                            column.getId(),
                            column.getName(),
                            totalTasks,
                            (int) doneTasks,
                            todoTasks,
                            Math.round(progress * 100.0) / 100.0
                    );
                })
                .collect(Collectors.toList());
    }

    public List<MemberStatisticsDto> getMemberStatistics(String projectId, String userId) {
        ensureUserIsProjectMember(projectId, userId);

        
        List<com.qlda.backendjava.columns.entity.ColumnEntity> columns = columnRepository.findByProjectId(projectId);
        List<String> columnIds = columns.stream()
                .map(com.qlda.backendjava.columns.entity.ColumnEntity::getId)
                .collect(Collectors.toList());

        
        List<TaskEntity> tasks = new ArrayList<>();
        for (String columnId : columnIds) {
            tasks.addAll(taskRepository.findByColumnIdOrderByPositionAsc(columnId));
        }

        Map<String, MemberStatisticsDto> userStatsMap = new HashMap<>();
        LocalDateTime now = LocalDateTime.now();

        for (TaskEntity task : tasks) {
            if (task.getAssignees() != null) {
                for (com.qlda.backendjava.users.entity.UserEntity assignee : task.getAssignees()) {
                    if (!userStatsMap.containsKey(assignee.getId())) {
                        userStatsMap.put(assignee.getId(), new MemberStatisticsDto(
                                assignee.getId(),
                                assignee.getName() != null ? assignee.getName() : assignee.getEmail(),
                                assignee.getAvatar() != null ? assignee.getAvatar() : "https://cdn-icons-png.flaticon.com/512/149/149071.png",
                                0, 0, 0, 0, 0.0
                        ));
                    }

                    MemberStatisticsDto stats = userStatsMap.get(assignee.getId());
                    stats.setTotalTasks(stats.getTotalTasks() + 1);

                    if (task.getStatus() == TaskEntity.Status.done) {
                        stats.setDoneTasks(stats.getDoneTasks() + 1);
                    } else {
                        stats.setTodoTasks(stats.getTodoTasks() + 1);
                        if (task.getDueDate() != null && task.getDueDate().isBefore(now)) {
                            stats.setOverdueTasks(stats.getOverdueTasks() + 1);
                        }
                    }
                }
            }
        }

        return userStatsMap.values().stream()
                .map(stats -> {
                    double completionRate = stats.getTotalTasks() == 0 ? 0 :
                            (stats.getDoneTasks() * 100.0 / stats.getTotalTasks());
                    stats.setCompletionRate(Math.round(completionRate * 100.0) / 100.0);
                    return stats;
                })
                .sorted((a, b) -> b.getTotalTasks() - a.getTotalTasks())
                .collect(Collectors.toList());
    }

    public List<TimelineStatisticsDto> getTimelineStatistics(String projectId, String period, 
                                                               LocalDateTime startDate, LocalDateTime endDate, String userId) {
        ensureUserIsProjectMember(projectId, userId);

        
        List<com.qlda.backendjava.columns.entity.ColumnEntity> columns = columnRepository.findByProjectId(projectId);
        List<String> columnIds = columns.stream()
                .map(com.qlda.backendjava.columns.entity.ColumnEntity::getId)
                .collect(Collectors.toList());

        
        List<TaskEntity> tasks = new ArrayList<>();
        for (String columnId : columnIds) {
            tasks.addAll(taskRepository.findByColumnIdOrderByPositionAsc(columnId));
        }

        Map<String, TimelineStatisticsDto> dateMap = new HashMap<>();

        for (TaskEntity task : tasks) {
            LocalDateTime createdDate = task.getCreatedAt();
            LocalDateTime completedDate = task.getCompletedAt();
            LocalDateTime dueDate = task.getDueDate();

            String createdKey = formatDateByPeriod(createdDate, period);
            String completedKey = completedDate != null ? formatDateByPeriod(completedDate, period) : null;

            
            if (!dateMap.containsKey(createdKey)) {
                dateMap.put(createdKey, new TimelineStatisticsDto(
                        createdKey, 0, 0, 0, 0
                ));
            }
            TimelineStatisticsDto createdStats = dateMap.get(createdKey);
            createdStats.setCreatedTasks(createdStats.getCreatedTasks() + 1);

            
            if (completedKey != null && task.getStatus() == TaskEntity.Status.done) {
                if (!dateMap.containsKey(completedKey)) {
                    dateMap.put(completedKey, new TimelineStatisticsDto(
                            completedKey, 0, 0, 0, 0
                    ));
                }
                TimelineStatisticsDto completedStats = dateMap.get(completedKey);
                completedStats.setCompletedTasks(completedStats.getCompletedTasks() + 1);

                
                if (dueDate != null && completedDate != null) {
                    if (!completedDate.isAfter(dueDate)) {
                        completedStats.setOnTimeTasks(completedStats.getOnTimeTasks() + 1);
                    } else {
                        completedStats.setLateTasks(completedStats.getLateTasks() + 1);
                    }
                }
            }
        }

        return dateMap.values().stream()
                .sorted(Comparator.comparing(TimelineStatisticsDto::getDate))
                .collect(Collectors.toList());
    }

    private String formatDateByPeriod(LocalDateTime date, String period) {
        if (period == null || period.equals("day")) {
            return date.toLocalDate().toString(); 
        } else if (period.equals("week")) {
            
            LocalDateTime weekStart = date.with(TemporalAdjusters.previousOrSame(DayOfWeek.SUNDAY));
            return weekStart.toLocalDate().toString(); 
        } else if (period.equals("month")) {
            
            return String.format("%d-%02d", date.getYear(), date.getMonthValue());
        } else {
            
            return date.toLocalDate().toString();
        }
    }

    public CommentStatisticsDto getCommentStatistics(String projectId, String filter, String userId) {
        ensureUserIsProjectMember(projectId, userId);

        
        List<com.qlda.backendjava.columns.entity.ColumnEntity> columns = columnRepository.findByProjectId(projectId);
        List<String> columnIds = columns.stream()
                .map(com.qlda.backendjava.columns.entity.ColumnEntity::getId)
                .collect(Collectors.toList());

        
        List<TaskEntity> tasks = new ArrayList<>();
        for (String columnId : columnIds) {
            tasks.addAll(taskRepository.findByColumnIdOrderByPositionAsc(columnId));
        }

        List<String> taskIds = tasks.stream()
                .map(TaskEntity::getId)
                .collect(Collectors.toList());

        if (taskIds.isEmpty()) {
            return new CommentStatisticsDto(0, 0, List.of(), List.of());
        }

        
        LocalDateTime startDate = null;
        if (filter != null && filter.equals("24h")) {
            startDate = LocalDateTime.now().minusDays(1);
        } else if (filter != null && filter.equals("7d")) {
            startDate = LocalDateTime.now().minusDays(7);
        }

        
        List<CommentEntity> comments = commentRepository.findByTaskIdsAndCreatedAtAfter(taskIds, startDate);

        int totalComments = comments.size();

        
        int recentComments;
        if (filter != null && filter.equals("24h")) {
            LocalDateTime yesterday = LocalDateTime.now().minusDays(1);
            recentComments = (int) comments.stream()
                    .filter(c -> c.getCreatedAt() != null && !c.getCreatedAt().isBefore(yesterday))
                    .count();
        } else if (filter != null && filter.equals("7d")) {
            LocalDateTime sevenDaysAgo = LocalDateTime.now().minusDays(7);
            recentComments = (int) comments.stream()
                    .filter(c -> c.getCreatedAt() != null && !c.getCreatedAt().isBefore(sevenDaysAgo))
                    .count();
        } else {
            recentComments = totalComments;
        }

        
        Map<String, CommentByTaskDto> commentsByTaskMap = new HashMap<>();
        for (CommentEntity comment : comments) {
            if (comment.getTask() != null) {
                String taskId = comment.getTask().getId();
                if (!commentsByTaskMap.containsKey(taskId)) {
                    commentsByTaskMap.put(taskId, new CommentByTaskDto(
                            taskId,
                            comment.getTask().getTitle() != null ? comment.getTask().getTitle() : "Untitled",
                            0
                    ));
                }
                CommentByTaskDto taskDto = commentsByTaskMap.get(taskId);
                taskDto.setCommentCount(taskDto.getCommentCount() + 1);
            }
        }

        
        Map<String, CommentByMemberDto> commentsByMemberMap = new HashMap<>();
        for (CommentEntity comment : comments) {
            if (comment.getUser() != null) {
                String commentUserId = comment.getUser().getId();
                if (!commentsByMemberMap.containsKey(commentUserId)) {
                    String userName = comment.getUser().getName() != null 
                            ? comment.getUser().getName() 
                            : comment.getUser().getEmail();
                    String avatar = comment.getUser().getAvatar() != null 
                            ? comment.getUser().getAvatar() 
                            : "https://cdn-icons-png.flaticon.com/512/149/149071.png";
                    commentsByMemberMap.put(commentUserId, new CommentByMemberDto(
                            commentUserId,
                            userName,
                            avatar,
                            0
                    ));
                }
                CommentByMemberDto memberDto = commentsByMemberMap.get(commentUserId);
                memberDto.setCommentCount(memberDto.getCommentCount() + 1);
            }
        }

        
        List<CommentByTaskDto> commentsByTask = commentsByTaskMap.values().stream()
                .sorted((a, b) -> b.getCommentCount().compareTo(a.getCommentCount()))
                .limit(10)
                .collect(Collectors.toList());

        List<CommentByMemberDto> commentsByMember = commentsByMemberMap.values().stream()
                .sorted((a, b) -> b.getCommentCount().compareTo(a.getCommentCount()))
                .limit(10)
                .collect(Collectors.toList());

        return new CommentStatisticsDto(
                totalComments,
                recentComments,
                commentsByTask,
                commentsByMember
        );
    }

    public DeadlineAnalyticsDto getDeadlineAnalytics(String projectId, String userId) {
        ensureUserIsProjectMember(projectId, userId);

        
        List<com.qlda.backendjava.columns.entity.ColumnEntity> columns = columnRepository.findByProjectId(projectId);
        List<String> columnIds = columns.stream()
                .map(com.qlda.backendjava.columns.entity.ColumnEntity::getId)
                .collect(Collectors.toList());

        
        List<TaskEntity> tasks = new ArrayList<>();
        for (String columnId : columnIds) {
            tasks.addAll(taskRepository.findByColumnIdOrderByPositionAsc(columnId).stream()
                    .filter(t -> t.getDueDate() != null)
                    .collect(Collectors.toList()));
        }

        LocalDateTime now = LocalDateTime.now();
        LocalDateTime threeDaysLater = now.plusDays(3);

        List<TaskDeadlineDto> overdueTasksList = new ArrayList<>();
        List<TaskDeadlineDto> dueSoonTasksList = new ArrayList<>();
        List<TaskDeadlineDto> completedOnTimeList = new ArrayList<>();
        List<TaskDeadlineDto> completedLateList = new ArrayList<>();

        for (TaskEntity task : tasks) {
            if (task.getDueDate() == null) continue;

            LocalDateTime dueDate = task.getDueDate();
            boolean isOverdue = dueDate.isBefore(now) && task.getStatus() != TaskEntity.Status.done;
            boolean isDueSoon = !dueDate.isBefore(now) && !dueDate.isAfter(threeDaysLater) 
                    && task.getStatus() != TaskEntity.Status.done;

            TaskDeadlineDto taskDto = new TaskDeadlineDto(
                    task.getId(),
                    task.getTitle(),
                    task.getDueDate(),
                    task.getStatus().name(),
                    task.getCompletedAt()
            );

            if (isOverdue) {
                overdueTasksList.add(taskDto);
            } else if (isDueSoon) {
                dueSoonTasksList.add(taskDto);
            } else if (task.getStatus() == TaskEntity.Status.done && task.getCompletedAt() != null) {
                LocalDateTime completedAt = task.getCompletedAt();
                if (!completedAt.isAfter(dueDate)) {
                    completedOnTimeList.add(taskDto);
                } else {
                    completedLateList.add(taskDto);
                }
            }
        }

        return new DeadlineAnalyticsDto(
                overdueTasksList.size(),
                dueSoonTasksList.size(),
                completedOnTimeList.size(),
                completedLateList.size(),
                overdueTasksList,
                dueSoonTasksList,
                completedOnTimeList,
                completedLateList
        );
    }

    
    private void ensureUserIsProjectMember(String projectId, String userId) {
        ProjectEntity project = projectRepository.findById(projectId)
                .orElseThrow(() -> new NotFoundException("Không tìm thấy dự án."));
        
        
        if (project.getOwner().getId().equals(userId)) {
            return;
        }
        
        
        boolean isMember = projectMemberRepository.existsByProjectIdAndUserId(projectId, userId);
        
        if (!isMember) {
            throw new ForbiddenException("Bạn không có quyền truy cập dự án này.");
        }
    }
}

